transports/httpclient/src/main/java/org/commonjava/maven/galley/transport/htcli/UploadMetadataGenTransferDecorator.java

```diff
@@ -48,96 +52,146 @@
 public class UploadMetadataGenTransferDecorator
         extends AbstractTransferDecorator
 {
+    private static final String HTTP_METADATA_WRITE = "io.http-metadata.write";
+    private static final String HTTP_METADATA_WRITE_OPEN = HTTP_METADATA_WRITE + ".open";
+
     private static final Logger logger = LoggerFactory.getLogger( UploadMetadataGenTransferDecorator.class );
 
     private SpecialPathManager specialPathManager;
 
-    public UploadMetadataGenTransferDecorator( SpecialPathManager specialPathManager )
+    private MetricRegistry metricRegistry;
+
+    public UploadMetadataGenTransferDecorator( SpecialPathManager specialPathManager, MetricRegistry metricRegistry )
     {
         this.specialPathManager = specialPathManager;
+        this.metricRegistry = metricRegistry;
     }
 
     @Override
     public OutputStream decorateWrite( OutputStream stream, Transfer transfer, TransferOperation op,
                                        EventMetadata metadata )
             throws IOException
     {
+        if ( transfer.getPath().endsWith( FILE_EXTENSION ) )
+        {
+            return super.decorateWrite( stream, transfer, op, metadata );
+        }
+
         final SpecialPathInfo specialPathInfo = specialPathManager.getSpecialPathInfo( transfer, metadata.getPackageType() );
 
         final Boolean isMetadata = specialPathInfo != null && specialPathInfo.isMetadata();
         final Boolean isUpload = op == TransferOperation.UPLOAD;
         final Boolean hasRequestHeaders =
                 metadata.get( STORE_HTTP_HEADERS ) != null && metadata.get( STORE_HTTP_HEADERS ) instanceof Map;
 
-        // We only care about non-metadata artifact uploading for http-metadata generation
         if ( isUpload && !isMetadata && hasRequestHeaders )
         {
-            @SuppressWarnings( "unchecked" ) Map<String, List<String>> storeHttpHeaders =
-                    (Map<String, List<String>>) metadata.get( STORE_HTTP_HEADERS );
-            writeMetadata( transfer, new ObjectMapper(), storeHttpHeaders );
+            // we defer writing http-metadata.json until the main stream closes, to avoid trouble with directory-level write locks.
+            return new HttpMetadataWrapperOutputStream( stream, transfer, metadata );
```

Comment:
Here seems the original stream is used. Currently is ok as AbstractTransferDecorator.decorateWrite directly return this original stream which is used, but I'm not sure if we will refactor it in future to add some other decorated logic. So maybe we should also pass the super.decorateWrite logic to this HttpMetadataWrapperOutputStream by some ways?